Python/Product/PythonTools/PythonTools/Debugger/DebugAdapter/DebugAdapterProcess.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Windows.Forms;
+using Microsoft.VisualStudio.Debugger.DebugAdapterHost.Interfaces;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudioTools.Infrastructure;
+
+namespace Microsoft.PythonTools.Debugger {
+    internal sealed class DebugAdapterProcess {
+        private readonly ITargetHostInterop _targetInterop;
+        private readonly string _debuggerAdapterDirectory;
+
+        private ITargetHostProcess _targetHostProcess;
+        private string _webBrowserUrl;
+
+        public DebugAdapterProcess(ITargetHostInterop targetInterop, string debuggerAdapterDirectory) {
+            _targetInterop = targetInterop ?? throw new ArgumentNullException(nameof(targetInterop));
+            _debuggerAdapterDirectory = debuggerAdapterDirectory ?? throw new ArgumentNullException(nameof(debuggerAdapterDirectory));
+        }
+
+        public ITargetHostProcess StartProcess(string pythonExePath, string webBrowserUrl) {
+            if (string.IsNullOrEmpty(pythonExePath)) {
+                MessageBox.Show(Strings.PythonInterpreterPathNullOrEmpty, Strings.ProductTitle, MessageBoxButtons.OK, MessageBoxIcon.Warning);
+                return null;
+            }
+
+            _webBrowserUrl = webBrowserUrl;
+            _targetHostProcess = _targetInterop.ExecuteCommandAsync(pythonExePath, "\"" + _debuggerAdapterDirectory + "\"");
+
+            if (!string.IsNullOrEmpty(webBrowserUrl) && Uri.TryCreate(webBrowserUrl, UriKind.RelativeOrAbsolute, out var uri)) {
+                OnPortOpenedHandler.CreateHandler(uri.Port, null, null, () => _targetHostProcess.HasExited, LaunchBrowserDebugger);
+            }
+
+            return _targetHostProcess;
+        }
+
+        private void LaunchBrowserDebugger() {
```

Comment:
>LaunchBrowserDebugger [](start = 21, length = 21)

This is currently done using `OnPortOpenedHandler`, once we have a way to listen to the messages this can be done entirely using that. VS Code uses pattern matching to extract the host+port from the output event.

Reply:
I'll keep it in mind